nucypher/blockchain/eth/clients.py

```diff
@@ -15,16 +16,109 @@
 )
 from geth.process import BaseGethProcess
 from twisted.logger import Logger
+from web3 import Web3
 
-from nucypher.config.constants import DEFAULT_CONFIG_ROOT, BASE_DIR, DEPLOY_DIR, USER_LOG_DIR
+from nucypher.config.constants import DEFAULT_CONFIG_ROOT, DEPLOY_DIR, USER_LOG_DIR
 
 NUCYPHER_CHAIN_IDS = {
     'devnet': 112358,
 }
 
 
-class NuCypherGethProcess(LoggingMixin, BaseGethProcess):
+class Web3ClientError(Exception):
+    pass
+
+
+class Web3ClientConnectionFailed(Web3ClientError):
+    pass
+
+
+class Web3ClientUnexpectedVersionString(Web3ClientError):
+    pass
+
+
+class Web3Client(object):
+
+    def __new__(self, w3: Web3, *args):
+        #
+        # *Client version format*
+        # Geth Example: "'Geth/v1.4.11-stable-fed692f6/darwin/go1.7'"
+        # Parity Example: "Parity//v1.5.0-unstable-9db3f38-20170103/x86_64-linux-gnu/rustc1.14.0"
+        # Ganache Example: "EthereumJS TestRPC/v2.1.5/ethereum-js"
+        #
+        client_data = w3.clientVersion.split('/')
+        node_technology = client_data[0]
+
+        GETH = 'Geth'
+        PARITY = 'Parity'
+        GANACHE = 'EthereumJS TestRPC'
+
+        try:
+            gclass = {
+                GETH: GethClient,
+                PARITY: ParityClient,
+                GANACHE: GanacheClient,
+            }[node_technology]
+            return gclass(w3, *client_data)
+
+        except KeyError:
+            raise NotImplementedError(node_technology)
+
+
+class Web3ClientBase(object):
+
+    def __init__(self, w3, node_technology, version, backend, **kwargs):
+        self.web3_instance = w3
+        self.node_technology = node_technology
+        self.node_version = version
+        self.backend = backend
+
+    @property
+    @abstractmethod
+    def peers(self):
+        raise NotImplementedError
+
+    @property
+    def syncing(self):
+        return self.web3_instance.eth.syncing
 
+    @abstractmethod
+    def unlock_account(self, address, password):
+        raise NotImplementedError
+
+    def is_connected(self):
+        return self.web3_instance.isConnected()
+
+
+class GethClient(Web3ClientBase):
+
+    @property
+    def peers(self):
+        return self.web3_instance.geth.admin.peers()
+
+    def unlock_account(self, address, password):
+        return self.web3_instance.geth.personal.unlockAccount(address, password)
+
+
+class ParityClient(Web3ClientBase):
+
+    def peers(self) -> str:
+        return self.web3_instance.manager.request_blocking("parity_netPeers", [])
+
+    def unlock_account(self, address, password):
+        return self.web3_instance.parity.unlockAccount.unlockAccount(address, password)
+
+
+class GanacheClient(Web3ClientBase):
+
+    def peers(self):
+        return []
```

Comment:
Are the return values for `peers` and `unlock_account` a result of how Ganache works or should we return a `NotImplementedError` instead?

Reply:
I just assumed “don’t crash the whole thing” in these cases rather than raise an exception.  Maybe we will want to.

Reply:
Maybe the thing to do is write the access to these methods in a way that these methods will never be hit if the provider doesn’t support it.

Reply:
There may be workarounds for those methods. Previously, I couldn't find an endpoint for `peers` but there was one for `unlock_account` - see `personal_unlockAccount` in https://github.com/trufflesuite/ganache-cli. It may not need to be implemented in this PR however - but something to think about.

@KPrasch ?

Reply:
It may be best to crash for now, until we figure out what to do. Purposely not hitting these methods could produce some ugly code 🤢

Reply:
What I mean is right now I think peers is called in sync() somewhere and in the case of ganache there is no syncing, so we could just implement sync on these client classes and ganacheclient would never call its peers() so it wouldn’t matter.  But yeah in that case raising is fine

Reply:
👍  - yeah implementing `sync` in the clients themselves and having it called from `Blockchain.sync()` would be good. Assuming Ganache does not require syncing by nature then sure the `sync` method for the GanacheClient would do nothing.

I just didn't want `Blockchain.sync()` to have a bunch of `if` statements checking client type.